py/server/deephaven/jcompat.py

```diff
@@ -255,26 +255,49 @@ def _j_array_to_numpy_array(dtype: DType, j_array: jpy.JType, conv_null: bool, t
         np_array = np.array(j_array, np.object_)
 
     if conv_null:
-        if dh_null := _PRIMITIVE_DTYPE_NULL_MAP.get(dtype):
-            if dtype in (dtypes.float32, dtypes.float64):
-                np_array = np.copy(np_array)
-                np_array[np_array == dh_null] = np.nan
-            else:
-                if dtype is dtypes.bool_:  # needs to change its type to byte for dh null detection
-                    np_array = np.frombuffer(np_array, np.byte)
-
-                if any(np_array[np_array == dh_null]):
-                    if not type_promotion:
-                        raise DHError(f"Problem creating numpy array.  Java {dtype} array contains Deephaven null values, but numpy {np_array.dtype} array does not support null values")
-                    np_array = np_array.astype(np.float64)
-                    np_array[np_array == dh_null] = np.nan
-                else:
-                    if dtype is dtypes.bool_:  # needs to change its type back to bool
-                        np_array = np.frombuffer(np_array, np.bool_)
-                    return np_array
+        return dh_null_to_nan(np_array, type_promotion)
 
     return np_array
 
+def dh_null_to_nan(np_array: np.ndarray, type_promotion: bool = True) -> np.ndarray:
```

Comment:
When I first looked at the name of this function, I thought it converted between DH (Java) Null and NaN. The docstring says NumPy NaN. Is there a reason we don't name this `dh_null_to_np_nan` to make it more obvious to the user what it's converting to?

Reply:
The docstring says Numpy NaN because we are converting numpy arrays and want to support both float32 and float64, but NaN itself is implemented the same way across the packages/languages as long as they adhere to IEEE 754, so I am not sure prefix it with np in the name is very helpful while making the name just a bit too long.

Reply:
I would not count on all users to understand that NaN is implemented in the same way across packages/languages. That said, I do think it's reasonable to expect that they read the documentation, especially if they're confused about usage. I'm ok with the shorter name.